vyper/venom/passes/algebraic_optimization.py

```diff
@@ -0,0 +1,66 @@
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.basicblock import IRInstruction, IROperand
+from vyper.venom.passes.base_pass import IRPass
+
+
+class AlgebraicOptimizationPass(IRPass):
+    """
+    This pass reduces algebraic evaluatable expressions.
+
+    It currently optimizes:
+        * iszero chains
+    """
+
+    def _optimize_iszero_chains(self) -> None:
+        fn = self.function
+        for bb in fn.get_basic_blocks():
+            for inst in bb.instructions:
+                if inst.opcode != "iszero":
+                    continue
+
+                iszero_chain = self._get_iszero_chain(inst.operands[0])
+                iszero_count = len(iszero_chain)
+                if iszero_count == 0:
+                    continue
+
+                for use_inst in self.dfg.get_uses(inst.output):
+                    opcode = use_inst.opcode
+
+                    if opcode == "iszero":
+                        # We keep iszero instuctions as is
+                        continue
+                    if opcode in ("jnz",):
+                        # instructions that accept an integer as input:
+                        # we can remove up to all the iszero instructions
+                        keep_count = 1 - iszero_count % 2
+                    else:
+                        # instructions that accept a boolean as input:
+                        # we need to keep at least one or two iszero instructions
+                        keep_count = 1 + iszero_count % 2
+
+                    if keep_count >= iszero_count:
+                        continue
+
+                    out_var = iszero_chain[keep_count].operands[0]
+                    use_inst.replace_operands({inst.output: out_var})
+
+    def _get_iszero_chain(self, op: IROperand) -> list[IRInstruction]:
+        chain: list[IRInstruction] = []
+
+        while True:
+            inst = self.dfg.get_producing_instruction(op)
+            if inst is None or inst.opcode != "iszero":
+                break
+            op = inst.operands[0]
+            chain.insert(0, inst)
```

Comment:
`chain.append(inst)` followed by `chain.reverse()` at the end can avoid some performance blowups for long lists